### PR TITLE
First commit - adding new check

### LINKED ISSUE #62: Add check: Check DS with multiple death records with death dates, where death dates do not match

### PR DESCRIPTION
**Pull request by fernank8**
Closes #62

### CODE REVIEW

File: check_ds_multdeath_dsstdtc.R [hunk 1]
```diff
@@ -0,0 +1,67 @@
...
+#'
+#' @examples
+#'
+#' DS_error <- data.frame(STUDYID = rep(1, 6),
+#'   USUBJID = c(1, 1, 1, 2, 1,1),
+#'   DSDECOD = c("DEATH", "DEATH", rep("", 2),"DEATH","DEATH"),
+#'   DSSCAT = LETTERS[1:6],
+#'   DSSTDTC = c("", "2016-01-01", "", "", "2016-01-02","2016-01-01"),
+#'   stringsAsFactors = FALSE)
+
```
[the remark this answers is not in the document]
Comment (fernank8) [reply check_ds_multdeath_dsstdtc.R]: Thank you! Addressed!

File: check_ds_multdeath_dsstdtc.R [hunk 2]
```diff
@@ -0,0 +1,67 @@
...
+#' DS_noerror <- data.frame(STUDYID = rep(1, 6),
+#'                         USUBJID = c(1, 1, 1, 2, 1,1),
+#'                         DSDECOD = c("DEATH", "DEATH", rep("", 2),"DEATH","DEATH"),
+#'                         DSSCAT = LETTERS[1:6],
+#'                         DSSTDTC = c("", "2016-01-01", "", "", "2016-01-01","2016-01-01"),
+#                         stringsAsFactors = FALSE)
```
[the remark this answers is not in the document]
Comment (fernank8) [reply check_ds_multdeath_dsstdtc.R]: Thanks for explaining this! Addressed!

**Comment by fernank8** [comment]
Thank you for the suggestions Sara! I have updated the description and fail message as per your advice. (I'm getting great practice pull/pushing, fantastic! :D 

Regarding your comment "This passes even though the other records have a date of 2016-01-01. How should partial missing death dates be handled when the month does not match? What about if just "2016-01" were reported?" thank you for testing this case!  I have updated the code so that it no longer checks if non-missing death dates DSSTDTC match, but rather whether non-missing death text DSSTDTC match.  So now in the example you gave, the function should return an error.

Looking forward to hearing what you think! Thank you so much!

**Comment by fernank8** [comment]
Thank you Sara for your update suggestions!  I have added an underscore to DS_error2 and have also updated the order of the examples. And glad you agree with my logic re: partial death dates :)  Please let me know if you suggest any other updates.  Thank you so much!!

**Comment by harriscw** [comment]
Thanks @fernank8 for writing and thanks @sarabodach for checking.  I'll take a look.

FYI I think as long as some function somewhere specifies a function is imported that is sufficient.  You can see here that `dplyr:n()` is already imported:

https://github.com/pharmaverse/sdtmchecks/blob/f1d1a802a054c75703960eadc85802301d95d454/NAMESPACE#L148

So while its nice to specify which external functions are used in `@importFrom` its seems its not strictly necessary if done elsewhere

**Comment by fernank8** [comment]
Thank you @sarabodach and @harriscw !

@sarabodach since group_by is also already imported by sdtmchecks, do you think I should still add it to the header?  Thank you for confirming :-)

**Comment by harriscw** [comment]
Should something like this be included? e.g. for Roche to parse DSSPID for easier record location?
 https://github.com/pharmaverse/sdtmchecks/blob/e505c04fce2f4dfb757168bb5f532bc2557700ee/R/check_dd_ae_aedthdtc_ds_dsstdtc.R#L68

**Comment by fernank8** [comment]
Thank you so much @harriscw for your review and updates!

Regarding what `length(unique(DSSTDTC))>1` is for:

- When I initially wrote the check, I used filter(n() >=2, all(abs(diff(as.Date(DSSTDTC))) > 0)) instead of  filter(n() >=2, length(unique(DSSTDTC))>1). The idea was that if there were multiple records with dates, I was going to compare all of the dates to ensure that they were the same.  If there were multiple dates but they were the same, then the check would pass.

- However, @sarabodach asked what the check should do if there were partial dates.  If there were partial dates, then the code I initially wrote would not work, because the as.Date function would not work for partial dates.

- @sarabodach and I agreed that in the case of partial dates, as long as the partial dates were the same, then the check should pass.

- So I updated  filter(n() >=2, all(abs(diff(as.Date(DSSTDTC))) > 0)) to filter(n() >=2, length(unique(DSSTDTC))>1) with the idea being that I check if there were multiple records with a "date" (which could include a partial date, or really any text), and all I wanted to do was check that if there were multiple "dates", then all of the "dates" were the same. I believe that  length(unique(DSSTDTC))>1) would be 1 if all of the "dates" are the same, and greater than 1 if any dates are different. And if any of the dates are different, then the check would fail.

What do you think?

Thanks a lot @harriscw for your review!!

**Comment by harriscw** [comment]
@fernank8 Gotcha thanks yeah this makes sense.  Right now the check returns all records where there's more than one unique death date, even if some of the dates match.  For example
image
I think thats fine though, hopefully this is rare enough that it doesn't clog up too mach space in the report.  I also added back in `DSSCAT` as an optional variable (not required, display it if it exists) to help better identify the form.

**Comment by fernank8** [comment]
Sweet, all sounds good, thanks Will!

**Pull request merged**